addons/im_livechat/models/mail_channel.py

```diff
@@ -191,3 +194,134 @@ def _email_livechat_transcript(self, email):
+            field_name = inputs_dict[message_id.chatbot_step_id.step_type]
+            if not values.get(field_name):
+                values[field_name] = message_id.chatbot_user_raw_answer
+
+        return values
+
+    def _get_user_email_phone(self):
+        """
+        helper method to get user's email and phone
+        """
+        user_input = self._fetch_user_input_values({
+            'question_email': 'email',
+            'question_phone': 'phone',
+        })
+        return user_input.get('email', False), user_input.get('phone', False)
+
+    def _chatbot_post_message(self, chatbot, body, text_to_html=True):
+        if text_to_html:
+            body = plaintext2html(body)
+
+        return self.with_context(mail_create_nosubscribe=True).message_post(
+            author_id=chatbot.operator_partner_id.id,
+            body=body,
+            message_type='comment',
+            subtype_xmlid='mail.mt_comment',
+        )
+
+    def _chatbot_process_step(self, step_id):
+        self.ensure_one()
+        if self.livechat_chatbot_current_step_id and step_id.sequence <= self.livechat_chatbot_current_step_id.sequence:
+            raise ValidationError(_('The new step sequence must be higher than the old one.'))
+
+        # We change the current step to the new step
+        self.livechat_chatbot_current_step_id = step_id
+
+        human_operator = False
+        if step_id.step_type == 'forward_operator':
+            human_operator = self.livechat_channel_id._get_random_operator() if self.livechat_channel_id else False
+            if human_operator:
+                self.with_user(self.env.ref('base.user_root')).add_members(
+                    human_operator.partner_id.ids,
+                    open_chat_window=True,
+                    post_joined_message=False)
+
+                if step_id.message:
+                    # first post the message of the step (if we have one)
+                    posted_message = self._chatbot_post_message(step_id.chatbot_id, step_id.message)
+                else:
+                    posted_message = False
+
+                # then post a small custom 'Operator has joined' notification
+                self._chatbot_post_message(
+                    step_id.chatbot_id,
+                    '<div class="o_mail_notification">%s</div>' % _('%s has joined', human_operator.partner_id.name),
+                    text_to_html=False)
+            else:
+                # don't post anything as the bot in case we did not find an operator
+                # the script will continue normally
+                posted_message = False
+        else:
+            posted_message = self._chatbot_post_message(step_id.chatbot_id, step_id.message)
+
+        if human_operator:
+            self._broadcast(human_operator.partner_id.ids)
+            self.channel_pin(self.uuid, pinned=True)
+
+        return posted_message
+
+    def _chatbot_validate_email(self, email_address, chatbot):
```

Comment:
Feels like as we have ``execute_step``, we could have ``check_step`` (or verify, assess, ...) who could be a somehow generic step-based method responsible of validating input. That way if we add the phone validation, it would be a simple addition still under a generic "check" method, easier to controllers or other code calling it.

Reply:
That seems cool, but maybe not necessary since phone_validation seems a bit complicated to do?
The validation also comes with specific code on the frontend side so I'm not sure we could have something completely generic.

Keeping the comment for now, will maybe have another look later if we have time.